benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/LinkGenerationBenchmarkBase.cs

```diff
@@ -0,0 +1,407 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing.Matching;
+using Microsoft.AspNetCore.Routing.Patterns;
+using Microsoft.AspNetCore.Routing.Template;
+using Microsoft.AspNetCore.Routing.Tree;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using Microsoft.Extensions.ObjectPool;
+
+namespace Microsoft.AspNetCore.Routing.LinkGeneration
+{
+    public abstract class LinkGenerationBenchmarkBase
+    {
+        protected IServiceProvider CreateServices(params Endpoint[] endpoints)
+        {
+            var services = new ServiceCollection();
+            services.AddSingleton<ObjectPoolProvider, DefaultObjectPoolProvider>();
+            services.AddLogging();
+            services.AddOptions();
+            services.AddRouting();
+
+            services.TryAddEnumerable(
+                ServiceDescriptor.Singleton<EndpointDataSource>(new DefaultEndpointDataSource(endpoints)));
+
+            return services.BuildServiceProvider();
+        }
+
+        protected void AssertUrl(string expectedUrl, string actualUrl)
+        {
+            AssertUrl(expectedUrl, actualUrl, StringComparison.Ordinal);
+        }
+
+        protected void AssertUrl(string expectedUrl, string actualUrl, StringComparison stringComparison)
```

Comment:
I like it. It is easy for micro-benchmarks to start doing the wrong thing and no one notices